### ISSUE TITLE
[clang] Static analyzer crash with core checkers

### ISSUE THREAD

**Issue by omern1** [llvm/llvm-project#117334]
https://gcc.godbolt.org/z/aMnzcs6Eb

``` C++
// > clang -O0 -Xclang -analyzer-checker=core -Xclang -analyze ./input.cpp

enum class a { b };
char c[sizeof(a)];
struct {
  operator a() { return __builtin_bit_cast(a, c); }
} d;
void e() {
  switch (d)
  case a::b:
}
```

**Comment by steakhal** [comment]
Confirmed.
```c++
enum class Enum : char { b };
Enum convert() {
    char c[1] = {};
    static_assert(sizeof(Enum) == 1);
    return __builtin_bit_cast(Enum, c);
}
void top() {
    switch (convert()) {
        case Enum::b:
            return;
    }
}
```

**Comment by steakhal** [comment]
Assertion `'assumeInclusiveRange' is not implementedfor this NonLoc`.
Stack trace:
```
 #8 SimpleConstraintManager::assumeInclusiveRangeInternal(ProgramStateRef, NonLoc, APSInt const&, APSInt const&, bool)
 #9 ConstraintManager::assumeDualImpl()
#10 ConstraintManager::assumeInclusiveRangeDual(ProgramStateRef, NonLoc, APSInt const&, APSInt const&)
#11 ExprEngine::processSwitch(SwitchNodeBuilder&)
#12 HandleBlockExit(clang::CFGBlock const*, ExplodedNode*)
```

**Comment by OCHyams** [comment]
I don't know anything about how the static analyzer works, but this similar test case crashes with a slightly different stack trace. I have no idea whether a fix for one would fix the other, so thought I'd note it just in case.
```
using a = int (*)(int);
char b[sizeof(a)];
void c() { __builtin_bit_cast(a, b)(int()); }
```
This one fails in `clang::ento::SimpleConstraintManager::assumeAux(class llvm::IntrusiveRefCntPtr<class clang::ento::ProgramState const>, class clang::ento::NonLoc, bool) SimpleConstraintManager.cpp:68`

**Comment by steakhal** [comment]
Well, personally I'm not particularly interested in fixing bitcast crashes. And I always wonder if these come from realworld code. If they would, then I could reconsider my position, but for now it's not too interesting.
Did you simplify this from fuzzing or from some real code?

**Comment by OCHyams** [comment]
> And I always wonder if these come from realworld code

These two cases are reduced from real world code, not from fuzzing

**Comment by wjristow** [comment]
> > And I always wonder if these come from realworld code
> 
> These two cases are reduced from real world code, not from fuzzing

I don't have experience working in the static analyzer, but this is still impacting us in the original realworld code. @steakhal , do you have any insight on this?

**Comment by NagyDonat** [comment]
Note that the `assumeInclusiveRange` function family is (in my opinion) a bit "cursed" -- it abuses the overflow behavior to express the $value \in range$ condition in a concise but counter-intuitive way. I hope that I can gradually eliminate it from the codebase by switching to other, more natural approaches.

(However, this is not immediately relevant for this bug -- my long-term plans are too slow to be a practical solution.)

**Comment by wjristow** [comment]
Thanks @NagyDonat for the update.  Given that a fix doesn't look like it's pending soon, we will work around this problem for now.

**Comment by ziqingluo-90** [comment]
Maybe related to: https://github.com/llvm/llvm-project/pull/139188

**Comment by wjristow** [comment]
> Maybe related to: [#139188](https://github.com/llvm/llvm-project/pull/139188)

Yes it is!  The fix for that (41229581a456619c341daca8354b8c9ae1b150df) fixes this issue as well.  Thanks for the fix @ziqingluo-90 , and thanks @OCHyams for pointing me at that fix.
